update slides and episodes for lesson 4 (#60)
carpentries-incubator/r-geospatial-urban: instructors/4-gis-slides.qmd

# Part 2. How to perform basic GIS operations with the `sf` package?

> ## Objectives:
> By the end of this session, you should be able to:  
> ## Objectives:
> By the end of this session, you should be able to:  
> 
> - Perform **geoprocessing operations** such as **union, join, and intersection**  
> - Create **buffers and centroids**  
> - Compute the **area** of spatial polygons  
> - Calculate **density** within spatial units  
> - **Map** the results  
> - Perform **geoprocessing operations** such as **union, join, and intersection**  
> - Create **buffers and centroids**  
> - Compute the **area** of spatial polygons  
> - Calculate **density** within spatial units  
> - **Map** the results  

Reviewer: ```suggestion
## Objectives:
By the end of this session, you should be able to:  

- Perform **geoprocessing operations** such as **union, join, and intersection**  
- Create **buffers and centroids**  
- Compute the **area** of spatial polygons  
- Calculate **density** within spatial units  
- **Map** the results 
```

Reviewer: Issue to be soleved in https://github.com/carpentries-incubator/r-geospatial-urban/issues/66

## the 'sf' package `r emo::ji("box")`

![](fig/sf_cheatsheet1.png){fig-align="center"}

## the 'sf' cheatsheet `r emo::ji("map")`

![](fig/sf_cheatsheet2.png){fig-align="center"}

## Conservation in Brielle, NL `r emo::ji("building")`

Let's focus on old buildings and imagine we're in charge of their
conservation. We want to know how much of the city would be affected by
a non-construction zone of 100m around pre-1800 buildings.

If you were using pen&paper or a traditional GIS software, what would your workflow be?

## Conservation in Brielle, NL `r emo::ji("building")`

```{r old-buildings}
bb <- osmdata::getbb("Brielle, NL")
x <- opq(bbox = bb) %>%
    add_osm_feature(key = 'building') %>%
    osmdata_sf()

buildings <- x$osm_polygons %>%
  st_transform(.,crs=28992)

summary(buildings$start_date)
```

## Conservation in Brielle, NL `r emo::ji("building")`

```{r }
old <- 1800 # year prior to which you consider a building old

buildings$start_date <- as.numeric(as.character(buildings$start_date))

old_buildings <- buildings %>%
  filter(start_date <= old)

ggplot(data = old_buildings) + 
   geom_sf(colour="red") +
   coord_sf(datum = st_crs(28992)) 
```

> ## Basic GIS operations

Reviewer 2: ```suggestion
## Basic GIS operations
As conservationists, we want to create a zone around historical
buildings where building regulation will have special restrictions to
preserve historical buildings.
## Buffer
Let's say this zone should be 100 meters. In GIS terms, we want to
create a buffer around polygons. The corresponding function `sf` is
`st_buffer`, with 2 arguments:
: the radius of the buffer.  

![Buffer example](https://saylordotorg.github.io/text_essentials-of-geographic-information-systems/section_11/5c7eb6fc4bcbfc5117d778f0aabfded6.jpg)  

## Buffer
Let's say this zone should be 100 meters. In GIS terms, we want to
create a buffer around polygons. The corresponding function `sf` is
`st_buffer`, with 2 arguments:
-   "x": the polygons around which to create buffers
-   "dist": the radius of the buffer.
```

```

## Basic GIS operations
As conservationists, we want to create a zone around historical
buildings where building regulation will have special restrictions to
preserve historical buildings.
## Buffer
Let's say this zone should be 100 meters. In GIS terms, we want to
create a buffer around polygons. The corresponding function `sf` is
`st_buffer`, with 2 arguments:
: the radius of the buffer.  

![Buffer example](https://saylordotorg.github.io/text_essentials-of-geographic-information-systems/section_11/5c7eb6fc4bcbfc5117d778f0aabfded6.jpg)  

## Buffer
Let's say this zone should be 100 meters. In GIS terms, we want to
create a buffer around polygons. The corresponding function `sf` is
`st_buffer`, with 2 arguments:
-   "x": the polygons around which to create buffers
-   "dist": the radius of the buffer.

As conservationists, we want to create a zone around historical

> buildings where building regulation will have special restrictions to

Reviewer: Something seems to be repeated here, like above. Can you please check?

Reviewer: Issue to be soleved in https://github.com/carpentries-incubator/r-geospatial-urban/issues/66

buffer_old_buildings <- 
  st_buffer(x = old_buildings, dist = distance)
 
ggplot(data = buffer_old_buildings) + 
  geom_sf() +   
  coord_sf(datum = st_crs(28992))

```

## Union

Now, we have overlapping buffers.

We would rather create a **unique** conservation zone rather than overlapping ones in that case. So we have to fuse the overlapping buffers into one polygon. This operation is called union and the corresponding function is st_union().

```{r union}
single_old_buffer <- st_union(buffer_old_buildings) %>%
   st_cast(to = "POLYGON") %>%

>   st_as_sf() 

Reviewer: A small detail... I see a few inconsistencies in the indentations throughout the presentation. Not a big issue on the slides, but it would be less confusing if the same two-space indentation is used for every level.

buffer_old_buildings <- 
  st_buffer(x = old_buildings, dist = distance)
 
ggplot(data = buffer_old_buildings) + 
  geom_sf() +   
  coord_sf(datum = st_crs(28992))

```

## Union

Now, we have overlapping buffers.

> We would rather create a **unique** conservation zone rather than overlapping ones in that case. So we have to fuse the overlapping buffers into one polygon. This operation is called union and the corresponding function is st_union().

Reviewer 2: ```suggestion
Instead of overlapping buffers, we want to create **continuous** conservation zones by merging adjacent areas into a single polygon.
This process is called a **union**, and the corresponding function in `sf` is `st_union()`.
**Union dissolves** boundaries between overlapping geometries.

```

  st_transform(.,crs=28992)  

ggplot() + 
  geom_sf(data = single_old_buffer, aes(fill=ID)) +
  geom_sf(data = centroids_old) +
  coord_sf(datum = st_crs(28992))
```

## Intersection

> Now, we would like to distinguish conservation areas based on the number of historic buildings they contain. In GIS terms, we would like to know how many centroids each fused buffer polygon contains. 

Reviewer 2: ```suggestion
Now, we would like to distinguish conservation areas based on the number of historic buildings they contain.
In GIS terms, we would like to know how many centroids each dissolved buffer polygon contains. 
```

> This operation means intersecting the layer of polygons with the layer of points and the corresponding function is `st_intersection()`.

Reviewer 2: ```suggestion
This operation means intersecting the polygon layer with the point layer.
The corresponding function is `st_intersection()`.
```

## Intersection

> `st_intersection` here adds the attributes of the intersected polygon buffers to the data table of the centroids. This means we will now know about each centroid, the ID of its intersected polygon-buffer, and a variable called “n” which is population with 1 for everyone. This means that all centroids will have the same weight when aggregated.

Reviewer 2: ```suggestion
`st_intersection` here adds the attributes of the intersected polygon buffers to the data table of the centroids. This means we will now know about each centroid, the ID of its intersected polygon-buffer, and a variable called “n” which is population with 1 for everyone. This means that all centroids will have the same weight when aggregated.

<div align="center">
    <img src="[IMAGE_URL_1](https://gisgeography.com/wp-content/uploads/2020/10/Intersect-Table-2048x1170.png)" alt="Point to Polygon" width="45%">
    <img src="https://gisgeography.com/wp-content/uploads/2020/10/Intersect-Tool.png" alt="Polygon to Polygon" width="45%">
</div>
```

```{r }
centroids_buffers <- 
  st_intersection(centroids_old,single_old_buffer) %>%
  mutate(n = 1)

```

## Intersection

We aggregate them by ID number (group_by(ID)) and sum the variable n to know how many centroids are contained in each polygon-buffer.

```{r}
centroid_by_buffer <- centroids_buffers %>%
   group_by(ID) %>%
   summarise(n = sum(n))
 
single_buffer <- single_old_buffer %>%
   mutate(n_buildings = centroid_by_buffer$n)
```

## Mapping the number of buildings

```{r} 
ggplot() + 
   geom_sf(data = single_buffer, aes(fill=n_buildings)) +
   scale_fill_viridis_c(alpha = 0.8,begin = 0.6,
                        end = 1, direction = -1,
                        option = "B")
```

Problem: there are many pre-war buildings and the buffers are large so the number of old buildings is not very meaningful. Let’s compute the density of old buildings per buffer zone.

> ## Area

Reviewer 2: ```suggestion
## Calculating area and density of spatial features
```

## Intersection

Now, we would like to distinguish conservation areas based on the number of historic buildings they contain.
In GIS terms, we would like to know how many centroids each dissolved buffer polygon contains. 

This operation means intersecting the polygon layer with the point layer.
The corresponding function is `st_intersection()`.

> The corresponding function is `st_intersection()`.

Reviewer: Repetition?

Reviewer: Issue to be soleved in https://github.com/carpentries-incubator/r-geospatial-urban/issues/66

  
single_buffer <- single_old_buffer %>% 
  mutate(n_buildings = centroid_by_buffer$n)
  
single_buffer$area <- st_area(single_buffer)  %>% 
  units::set_units(., km^2)
single_buffer$old_buildings_per_km2 <- as.numeric(single_buffer$n_buildings / single_buffer$area)

p <- ggplot() + 
   geom_sf(data = buildings) +
   geom_sf(data = single_buffer, 
           aes(fill=old_buildings_per_km2), colour = NA) +
   scale_fill_viridis_c(alpha = 0.6,
                        begin = 0.6,
                        end = 1,
                        direction = -1,
                        option = "B") 
  
```

## One solution

```{r}
p
```

## Summary and keypoints

> We have seen how to create spatial buffers and centroids, how to

Reviewer 2: ```suggestion
We have seen how to create **spatial buffers** and **centroids**, how to merge overlapping geometries using **union**, how to **intersect** vector data, and how to **retrieve the area** of polygons. Additionally, we have explored how to **calculate density**
```

:::
:::

## Thank you

-   the **workshop team**: Aleks, Ana, Claudiu, Clémentine, Daniele,

>     Javi, Jérôme, Kyri, Manuel, Paula, Selin + TU Library & BK Faculty

Reviewer: Make sure everyone is included here according to the website: https://4turesearchdata-carpentries.github.io/2025-02-10-tudelft/
I will also make sure on the intro slides that everyone is there.
